### ISSUE TITLE
Client route 404 flash

### ISSUE THREAD

**Issue by chrisfinch** [jonelantha/gatsby-s3-action#11]
Hi @jonelantha  - thanks for creating this action - really useful work.

Wondering if you can help me with an issue;

I am deploying my gatsby site to s3 using your action. I have a cloudfront distribution set up infront of the s3 bucket and invalidation from your action is working well. I have the whole thing behind a custom route53 domain.

I have gatsby set up with many client routes like so;
```
Route: mydomain.com/index.html
Route: mydomain.com/404.html
Route: mydomain.com/app
ClientRoute: mydomain.com/app/dashboard
ClientRoute: mydomain.com/app/users
More client routes ClientRoute: mydomain.com/app/<etc... >/<etc... >
```

The app route bootstraps the client app that handles the client route navigation.

I am having an issue where when I navigate from `/` to `/app/dashboard` after a successful login I get a flash of the 404 page before the client route kicks in and displays the correct page.

This also occurs on a hard refresh to a client route.

This issue is the same as the one described here: https://stackoverflow.com/a/52770491/849779 

I have tried to fix this using s3 redirects as described on stack overflow.

I have also set up a Lambda@edge function as you described in your page here: https://blog.elantha.com/cloudfront-index-lambda/

Unfortunately nothing seems to work. Wondering if you could perhaps share any other resources that could help me solve this issue?

Thanks in advance.

**Comment by jonelantha** [comment]
Hi @chrisfinch thanks for reaching out about this.
Yes I can appreciate the issue.

First thing would be to check that whatever navigates to the dashboard is using the Gatsby `<Link>` or `navigate` functions rather than a traditional link (see [Gatsby docs](https://www.gatsbyjs.org/docs/gatsby-link/)).  This may help with the flash of the 404 you see when navigating to the app.

I know the S3 redirects (as mentioned in that StackOverflow post) won't work if you setup CloudFront like I mentioned. Instead the redirects needs to be done in the lambda (this should also deal with the hard refresh issue). 

Basically the code in the lambda would have some additions to redirect any url starting with `/app/` to `/app/index.html` (or maybe `/app/index.html`?). I'll come back here with some suggestions of how to amend the lambda.

Closing this issue as it's not an issue with this component but please feel free to comment further here 😄

**Comment by chrisfinch** [comment]
@jonelantha Thanks so much for this reply, really helpful so far.

> First thing would be to check that whatever navigates to the dashboard is using the Gatsby <Link> or navigate functions rather than a traditional link (see Gatsby docs). This may help with the flash of the 404 you see when navigating to the app.

This definitely helped - I was navigating to the app route via `window.location` and switching that to the gatsby `navigate` eliminated the 404 flash for that route change.

However the issue still remains when you do a hard refresh or direct load of an app route with a dynamic route attached eg: `mydomain.com/app/dashboard`. I understand why this happens and as you said it would seem that updating the Lambda will be the correct thing to do.

I will have a go at that this week and update here if I have some success - perhaps you would like to update your documentation with a code sample to solve this issue as I couldn't find anything else online about it.

If you have any further suggestions I would like to hear them.

Thanks again 👌🙏🏻

**Comment by jonelantha** [comment]
Hi @chrisfinch 
I've looked into this a bit now. There's one solution which _may_ work:

You may be able to alter the lambda so it look like this:

```js
...
    if (looksLikeADirectory(requestPath) && !requestPath.endsWith('/')) {
        const pathWithSlash = `${requestPath}/`;

        return makeRedirect301Response(pathWithSlash, request.querystring);
    }
    
    if (requestPath.startsWith('/app')) {
        return makeOriginRequestWithNewPath(request, '/app/index.html');
    }

    if (requestPath.endsWith('/')) {
        const pathWithIndexHtml = `${requestPath}index.html`;

        return makeOriginRequestWithNewPath(request, pathWithIndexHtml);
    }
...    
```

(So inserting 3 lines into the existing one)

There is one catch: this may or may not be a dealbreaker for you - basically after this change, all files inside `/app` will be served as `/app/index.html`. If you've only got one `index.html` file inside `/app` then you'll be fine but otherwise you'll have no way to access those other files.

To make the change, find the lambda in the console, make sure you're editing the `$LATEST` version of the lambda (versions menu if you can see it, otherwise don't worry), when you've edited click save, then `deploy to lambd@edge` from the `actions` menu then on the popup make sure you say to use an existing trigger and make sure it's pointing at the right distribution.

Hope that helps 👍

**Comment by chrisfinch** [comment]
@jonelantha thanks so much for the suggestion - I finally got around to trying it out today.

Great news is that it worked with one small adjustment;

I had to add a trailing slash to app in `startsWith('/app')` so it became `startsWith('/app/')`.

This was because gatsby had output a JS file called `app-<HASH>.js` and the request for this file was also being redirected by the lambda which broke the app. Adding the trailing slash meant that the redirect now only applied to full paths starting with `app`.

Thanks again for the help as this was really useful. perhaps you might also want to update your documentation with the above as it could help out others.

And also thanks for making this open source code on behalf of the community - I appreciate you do this on your own time 👌
